Spigot-API-Patches/0297-More-Enchantment-API.patch

```diff
@@ -0,0 +1,124 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Thu, 6 May 2021 19:58:03 -0700
+Subject: [PATCH] More Enchantment API
+
+
+diff --git a/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java b/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e6a40c1fcea761bd66743b50e3da3d14797d05b0
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java
+@@ -0,0 +1,24 @@
++package io.papermc.paper.enchantments;
++
++public enum EnchantmentRarity {
++
++    COMMON(10),
++    UNCOMMON(5),
++    RARE(2),
++    VERY_RARE(1);
++
++    private final int weight;
++
++    EnchantmentRarity(int weight) {
++        this.weight = weight;
++    }
++
++    /**
++     * Gets the weight for the rarity.
++     *
++     * @return the weight
++     */
++    public int getWeight() {
++        return weight;
++    }
++}
+diff --git a/src/main/java/org/bukkit/enchantments/Enchantment.java b/src/main/java/org/bukkit/enchantments/Enchantment.java
+index 8eb0497c81744874809ebc4bc2e28b128e66a926..799813410f40589ba1d7d530a47d87f609764705 100644
+--- a/src/main/java/org/bukkit/enchantments/Enchantment.java
++++ b/src/main/java/org/bukkit/enchantments/Enchantment.java
+@@ -306,6 +306,46 @@ public abstract class Enchantment implements Keyed {
+      * @return the name of the enchantment with {@code level} applied
+      */
+     public abstract @NotNull net.kyori.adventure.text.Component displayName(int level);
++
++    /**
++     * Checks if this enchantment can be found in villager trades.
++     *
++     * @return true if the enchantment can be found in trades
++     */
++    public abstract boolean isTradeable();
++
++    /**
++     * Checks if this enchantment can be found in an enchanting table
++     * or use to enchant items generated by loot tables.
++     *
++     * @return true if the enchantment can be found in a table or by loot tables
++     */
++    public abstract boolean isDiscoverable();
++
++    /**
++     * Gets the rarity of this enchantment.
++     *
++     * @return the rarity
++     */
++    @NotNull
++    public abstract io.papermc.paper.enchantments.EnchantmentRarity getRarity();
++
++    /**
++     * Gets the damage increase as a result of the level and entity category specified
++     *
++     * @param level the level of enchantment
++     * @param entityCategory the category of entity
++     * @return the damage increase
++     */
++    public abstract float getDamageIncrease(int level, @NotNull org.bukkit.entity.EntityCategory entityCategory);
++
++    /**
++     * Gets the equipment slots where this enchantment is considered "active".
++     *
++     * @return the equipment slots
++     */
++    @NotNull
++    public abstract java.util.Set<org.bukkit.inventory.EquipmentSlot> getActiveSlots();
```

Comment:
Can't wait for people to be angry over these in their implementations 😆

Reply:
Should new methods not be abstract and just return null or 0 or an empty set or whatever and have them overriden in the wrapper and CraftEnchantment?

Reply:
No, absolutely not, they're fine as-is. Was mostly mocking folks who implement API classes that should never be extended, and they go out of their way to hacking in registrations.

Reply:
Yeah, that's what I figured, but thought I'd check to be sure.